deckhouse-controller/pkg/controller/module-controllers/release/updater.go

```diff
@@ -146,30 +153,39 @@ func (k *kubeAPI) DeployRelease(ctx context.Context, release *v1alpha1.ModuleRel
 		return fmt.Errorf("update module release download statistic: %w", err)
 	}
 
-	moduleVersionPath := path.Join(k.downloadedModulesDir, moduleName, "v"+release.Spec.Version.String())
+	tmpModuleVersionPath := path.Join(tmpDir, moduleName, "v"+release.Spec.Version.String())
 	relativeModulePath := generateModulePath(moduleName, release.Spec.Version.String())
-	newModuleSymlink := path.Join(k.symlinksDir, fmt.Sprintf("%d-%s", release.Spec.Weight, moduleName))
 
 	def := models.DeckhouseModuleDefinition{
 		Name:   moduleName,
 		Weight: release.Spec.Weight,
-		Path:   moduleVersionPath,
+		Path:   tmpModuleVersionPath,
 	}
-	err = validateModule(def)
+	var values = make(addonutils.Values)
+	if module := k.moduleManager.GetModule(moduleName); module != nil {
+		values = module.GetConfigValues(false)
+	}
+	err = validateModule(def, values)
 	if err != nil {
-		k.logger.Errorf("Module '%s:v%s' validation failed: %s", moduleName, release.Spec.Version.String(), err)
 		release.Status.Phase = v1alpha1.PhaseSuspended
-		if e := k.UpdateReleaseStatus(release, "validation failed: "+err.Error(), release.Status.Phase); e != nil {
-			return e
-		}
+		_ = k.UpdateReleaseStatus(release, "validation failed: "+err.Error(), release.Status.Phase)
+		return fmt.Errorf("module '%s:v%s' validation failed: %s", moduleName, release.Spec.Version.String(), err)
+	}
 
-		return nil
+	moduleVersionPath := path.Join(k.downloadedModulesDir, moduleName, "v"+release.Spec.Version.String())
+	if err = os.RemoveAll(moduleVersionPath); err != nil {
```

Comment:
what's it? Why remove anything from /deckhouse/downloaded/<module_name> directory?
Frankly speaking, I didn't get the idea of using tmpDir instead of downloading right to /deckhouse/downloaded/. Could you please elaborate it a bit?

Reply:
If you look at how we load the module into the file system now, you can see that we first delete the existing directory. And only after that we unpack it into the newly created directory.

[deckhouse-controller/pkg/controller/module-controllers/downloader/module_downloader.go](https://github.com/deckhouse/deckhouse/blob/main/deckhouse-controller/pkg/controller/module-controllers/downloader/module_downloader.go#L184)

This is done in order to correctly update the existing version and not get a module with the remaining files from the previous version variation.

Since we switched to using a temporary directory, this logic had to be pulled higher, from the function here.

Reply:
it's rather abnormal case when we already have a directory for x.y.z version of a module and the same version of module release is applied again. I suppose, that cleanup was added just to make sure we wouldn't have any intersections. Still can't wrap my head around the idea of using tmp directories.

Reply:
Are there any other ideas for implementing module validation without using temporary directories?

Reply:
Using tmp directories in general is fine, I don't mind, just wanted to wrap my head around it, thanks.